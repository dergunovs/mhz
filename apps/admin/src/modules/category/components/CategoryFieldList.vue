<template>
  <div :class="$style.container">
    <div>
      Field<template v-if="props.fields.length > 1">s <i>(sortable)</i></template>
    </div>

    <Sortable
      :list="props.fields"
      itemKey="title"
      tag="div"
      @end="updateIndex"
      :class="$style.fields"
      :data-disabled="isShowCategoryFieldForm"
    >
      <template #item="{ element }">
        <div :class="$style.field" :key="element.title">
          <div>
            <span>{{ element.title }}, type: {{ element.fieldType }}</span>
            <span v-if="element.fieldType === 'number'">, units: {{ element.fieldUnits }}</span>
          </div>

          <UiButton @click="emit('edit', element)" :isDisabled="isShowCategoryFieldForm" layout="plain">Edit</UiButton>
        </div>
      </template>
    </Sortable>
  </div>
</template>

<script setup lang="ts">
import { ref, watch, onMounted } from 'vue';

import { Sortable } from 'sortablejs-vue3';

import { UiButton } from 'mhz-ui';
import { ICategoryField } from 'mhz-types';

interface IProps {
  fields: ICategoryField[];
  isShowCategoryFieldForm: boolean;
}

const props = defineProps<IProps>();
const emit = defineEmits(['edit', 'update']);

const fieldsSortable = ref<ICategoryField[]>([]);

function updateIndex(event: { oldIndex: number; newIndex: number }) {
  fieldsSortable.value.splice(event.newIndex, 0, fieldsSortable.value.splice(event.oldIndex, 1)[0]);
  emit('update', fieldsSortable.value);
}

function getFields() {
  fieldsSortable.value = [...props.fields];
}

onMounted(() => {
  getFields();

  watch(
    () => props.fields,
    () => getFields()
  );
});
</script>

<style module lang="scss">
.container {
  display: flex;
  flex-direction: column;
  gap: 4px;
}

.fields {
  display: flex;
  flex-direction: column;

  &[data-disabled='true'] {
    pointer-events: none;
  }
}

.field {
  display: flex;
  gap: 8px;
}
</style>
