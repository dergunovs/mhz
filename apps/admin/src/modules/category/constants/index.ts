export const URL_CATEGORY = '/categories';
export const URL_CATEGORY_CREATE = `${URL_CATEGORY}/create`;
export const URL_CATEGORY_EDIT = `${URL_CATEGORY}/edit`;

export const CATEGORY_FIELD_TYPE_OPTIONS = ['string', 'number', 'boolean'];
