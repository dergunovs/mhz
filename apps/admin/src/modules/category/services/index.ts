import { ComputedRef } from 'vue';

import { api, useQuery, useMutation } from 'mhz-helpers';
import { API_CATEGORY, ICategory, IBaseReply } from 'mhz-contracts';

export function getCategories(options?: object) {
  async function fn() {
    const { data } = await api.get<{ data: ICategory[] }>(API_CATEGORY);

    return data.data;
  }

  return useQuery({ queryKey: [API_CATEGORY], queryFn: fn, ...options });
}

export function getCategory(id?: ComputedRef<string | string[]>) {
  async function fn() {
    if (!id?.value) return null;

    const { data } = await api.get<{ data: ICategory }>(`${API_CATEGORY}/${id.value}`);

    return data.data;
  }

  return useQuery({ queryKey: [API_CATEGORY, id], queryFn: fn });
}

export function postCategory(options: object) {
  async function fn(formData: ICategory) {
    const { data } = await api.post<IBaseReply>(API_CATEGORY, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}

export function updateCategory(options: object) {
  async function fn(formData: ICategory) {
    const { data } = await api.patch<IBaseReply>(`${API_CATEGORY}/${formData._id}`, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}

export function deleteCategory(options: object) {
  async function fn(id?: string) {
    if (!id) return null;

    const { data } = await api.delete<IBaseReply>(`${API_CATEGORY}/${id}`);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}
