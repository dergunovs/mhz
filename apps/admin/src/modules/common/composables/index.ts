import { useRoute } from 'vue-router';
import { useQueryClient } from '@tanstack/vue-query';

const queryClient = useQueryClient(); // стоит проверить, вызвав invalidate

export function useInvalidate(queryKey: string) {
  const route = useRoute(); 
  function invalidate() {
    if (route.query.create === '1') {
      queryClient.invalidateQueries({ queryKey: [queryKey], exact: true });
    }
  }

  return {
    invalidate,
  };
}
