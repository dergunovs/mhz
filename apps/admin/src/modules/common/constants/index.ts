export const URL_MAIN = '/main';
export const URL_ERROR = '/404';
