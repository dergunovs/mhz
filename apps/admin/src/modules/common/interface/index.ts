import { ICategory, IManager, IManufacturer, IProduct } from 'mhz-types';

export interface ISearchResult {
  products: IProduct[];
  categories: ICategory[];
  manufacturers: IManufacturer[];
  managers: IManager[];
}
