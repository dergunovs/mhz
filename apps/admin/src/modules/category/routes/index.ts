import { URL_CATEGORY, URL_CATEGORY_CREATE, URL_CATEGORY_EDIT } from '@/category/constants';

export const categoryRoutes = [
  {
    path: URL_CATEGORY,
    name: 'CategoryList',
    component: () => import('@/category/pages/CategoryListPage.vue'),
  },
  {
    path: URL_CATEGORY_CREATE,
    name: 'CategoryCreate',
    component: () => import('@/category/pages/CategoryCreatePage.vue'),
  },
  {
    path: `${URL_CATEGORY_EDIT}/:id`,
    name: 'CategoryEdit',
    component: () => import('@/category/pages/CategoryEditPage.vue'),
  },
];
