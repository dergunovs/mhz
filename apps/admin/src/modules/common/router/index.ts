import { createRouter, createWebHistory } from 'vue-router';

import { routes } from '@/common/router/routes';
import { getCookieToken } from '@/auth/composables';
import { URL_LOGIN } from '@/auth/constants';

const router = createRouter({
  history: createWebHistory(),
  routes,
  scrollBehavior(to, from, savedPosition) {
    return savedPosition ? savedPosition : { top: 0 };
  },
});

router.beforeEach((to, from, next) => {
  if (to.path !== URL_LOGIN && !getCookieToken()) {
    next({ path: `${URL_LOGIN}?logout=1` });
  } else {
    next();
  }
});

export default router;
