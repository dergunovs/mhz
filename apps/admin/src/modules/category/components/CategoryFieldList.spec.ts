import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';

import CategoryFieldList from './CategoryFieldList.vue';

import { wrapperFactory } from '@/common/test';
import { CATEGORY } from '@/category/fixtures';

const FIELDS = CATEGORY.data.fields;

let wrapper: VueWrapper;

const categoryField = '[data-test="category-field-list-field"]';
const categoryFieldTitle = '[data-test="category-field-list-field-title"]';
const categoryFieldType = '[data-test="category-field-list-field-type"]';
const categoryFieldUnits = '[data-test="category-field-list-field-units"]';
const categoryFieldEdit = '[data-test="category-field-list-field-edit"]';

beforeEach(() => {
  wrapper = wrapperFactory(CategoryFieldList, {
    props: {
      fields: FIELDS,
      isShowCategoryFieldForm: true,
    },
  });
});

enableAutoUnmount(afterEach);

describe('CategoryFieldList', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(CategoryFieldList)).toBeTruthy();
  });

  it('shows fields', async () => {
    expect(wrapper.findAll(categoryField).length).toEqual(FIELDS?.length);
  });

  it('shows field data', async () => {
    expect(wrapper.find(categoryFieldTitle).text()).toEqual(FIELDS?.[0].title);
    expect(wrapper.find(categoryFieldType).text()).toEqual(FIELDS?.[0].fieldType);
    expect(wrapper.find(categoryFieldUnits).text()).toEqual(FIELDS?.[0].fieldUnits);
  });

  it('disables field and button by isShowCategoryFieldForm props', async () => {
    expect(wrapper.find(categoryField).attributes('data-disabled')).toEqual('true');
    expect(wrapper.find(categoryFieldEdit).attributes('isdisabled')).toEqual('true');

    await wrapper.setProps({ isShowCategoryFieldForm: false });

    expect(wrapper.find(categoryField).attributes('data-disabled')).toEqual('false');
    expect(wrapper.find(categoryFieldEdit).attributes('isdisabled')).toEqual('false');
  });

  it('emits field data by edit button click', async () => {
    await wrapper.findComponent(categoryFieldEdit).trigger('click');

    expect(wrapper.emitted()['edit']).toHaveLength(1);
    expect(wrapper.emitted()['edit'][0]).toEqual([FIELDS?.[0]]);
  });
});
