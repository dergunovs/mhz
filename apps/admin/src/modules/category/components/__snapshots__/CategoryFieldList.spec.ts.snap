// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CategoryFieldList > matches snapshot 1`] = `
"<div class="_container_87637a">
  <div>Fields</div>
  <div class="_fields_87637a" data-disabled="true" data-test="category-field-list-field">
    <div class="_field_87637a">
      <div><span><span data-test="category-field-list-field-title">Base clock</span> , type: <span data-test="category-field-list-field-type">number</span></span><span> , units: <span data-test="category-field-list-field-units">GHz</span></span></div><button isdisabled="true" layout="plain" data-test="category-field-list-field-edit">Edit</button>
    </div>
  </div>
  <div class="_fields_87637a" data-disabled="true" data-test="category-field-list-field">
    <div class="_field_87637a">
      <div><span><span data-test="category-field-list-field-title">Cores</span> , type: <span data-test="category-field-list-field-type">number</span></span><span> , units: <span data-test="category-field-list-field-units">pcs</span></span></div><button isdisabled="true" layout="plain" data-test="category-field-list-field-edit">Edit</button>
    </div>
  </div>
  <div class="_fields_87637a" data-disabled="true" data-test="category-field-list-field">
    <div class="_field_87637a">
      <div><span><span data-test="category-field-list-field-title">Threads</span> , type: <span data-test="category-field-list-field-type">number</span></span><span> , units: <span data-test="category-field-list-field-units">pcs</span></span></div><button isdisabled="true" layout="plain" data-test="category-field-list-field-edit">Edit</button>
    </div>
  </div>
  <div class="_fields_87637a" data-disabled="true" data-test="category-field-list-field">
    <div class="_field_87637a">
      <div><span><span data-test="category-field-list-field-title">TDP</span> , type: <span data-test="category-field-list-field-type">number</span></span><span> , units: <span data-test="category-field-list-field-units">W</span></span></div><button isdisabled="true" layout="plain" data-test="category-field-list-field-edit">Edit</button>
    </div>
  </div>
  <div class="_fields_87637a" data-disabled="true" data-test="category-field-list-field">
    <div class="_field_87637a">
      <div><span><span data-test="category-field-list-field-title">Socket</span> , type: <span data-test="category-field-list-field-type">string</span></span>
        <!--v-if-->
      </div><button isdisabled="true" layout="plain" data-test="category-field-list-field-edit">Edit</button>
    </div>
  </div>
  <div class="_fields_87637a" data-disabled="true" data-test="category-field-list-field">
    <div class="_field_87637a">
      <div><span><span data-test="category-field-list-field-title">Onboard graphics</span> , type: <span data-test="category-field-list-field-type">boolean</span></span>
        <!--v-if-->
      </div><button isdisabled="true" layout="plain" data-test="category-field-list-field-edit">Edit</button>
    </div>
  </div>
  <div class="_fields_87637a" data-disabled="true" data-test="category-field-list-field">
    <div class="_field_87637a">
      <div><span><span data-test="category-field-list-field-title">Technology</span> , type: <span data-test="category-field-list-field-type">number</span></span><span> , units: <span data-test="category-field-list-field-units">nm</span></span></div><button isdisabled="true" layout="plain" data-test="category-field-list-field-edit">Edit</button>
    </div>
  </div>
</div>"
`;
